### DISCUSSION TITLE
Gitea not work correctly through reproxy.

### DISCUSSION THREAD

**Discussion by saaremaa** [umputun/reproxy#96] [Q&A]
Gitea version 1.14.3 built with GNU Make 4.1, go1.16.5 : bindata, sqlite, sqlite_unlock_notify
reproxy v0.8.0-4-g4fce9c4-master-20210620-20:43:04
settings /etc/reproxy
`git.mydomain.com:
  - {route: "^/(.*)", dest: "http://localhost:3000/$1"}`

I can't see commits in Gitea if it work through reproxy(502 returns), but i can if get comit directly. Can someone confirm this bug or i have a problem with settings?

**Comment by umputun** [comment]
are you committing over http(s) or ssh? is it running directly on the host or inside of the docker container? what the error message looks like on the git side?

**Reply by saaremaa** [reply]
Gitea is running on the server, not in docker. The problem only when working with the Gitea via the web interface. HTTPS  through reproxy, and HTTP directly.

Gitea-log:
2021/06/20 21:19:31 Completed GET /serviceworker.js 200 OK in 5.051148ms
2021/06/20 21:20:00 Completed GET /user/events 200 OK in 30.001884027s
2021/06/20 21:20:02 Started GET /saaremaa/przd-90/commit/899c64ce036140f9e22af136f13c2678172cc4c9 for [deleted]:0
2021/06/20 21:20:03 Started GET /serviceworker.js for [deleted]:0
2021/06/20 21:20:03 ...les/public/public.go:165:handle() [I] [Static] Serving /serviceworker.js
2021/06/20 21:20:03 Completed GET /serviceworker.js 304 Not Modified in 6.617503ms
2021/06/20 21:20:08 Started GET /serviceworker.js for [deleted]:0
2021/06/20 21:20:08 ...les/public/public.go:165:handle() [I] [Static] Serving /serviceworker.js
2021/06/20 21:20:08 Completed GET /serviceworker.js 304 Not Modified in 4.394584ms
2021/06/20 21:20:09 Completed GET /saaremaa/przd-90/commit/899c64ce036140f9e22af136f13c2678172cc4c9 200 OK in 6.987327794s
2021/06/20 21:20:14 Started GET /api/v1/repos/search?sort=updated&order=desc&uid=1&team_id=0&q=&page=1&limit=15&mode=&archived=false for [2a03:580
2021/06/20 21:20:14 Started GET /api/v1/repos/search?sort=updated&order=desc&uid=1&team_id=0&q=&page=1&mode= for [2a03:5800:f8f8:e600:a5ce:2e22:23
2021/06/20 21:20:14 Started GET /user/events for [deleted]:0
2021/06/20 21:20:14 Completed GET /api/v1/repos/search?sort=updated&order=desc&uid=1&team_id=0&q=&page=1&limit=15&mode=&archived=false 200 OK in 4
2021/06/20 21:20:14 Completed GET /api/v1/repos/search?sort=updated&order=desc&uid=1&team_id=0&q=&page=1&mode= 200 OK in 49.592879ms
2021/06/20 21:20:15 Started GET /serviceworker.js for [deleted]:0
2021/06/20 21:20:15 ...les/public/public.go:165:handle() [I] [Static] Serving /serviceworker.js
2021/06/20 21:20:15 Completed GET /serviceworker.js 304 Not Modified in 3.69672ms
2021/06/20 21:20:21 Started GET /serviceworker.js for [deleted]:0
2021/06/20 21:20:21 ...les/public/public.go:165:handle() [I] [Static] Serving /serviceworker.js
2021/06/20 21:20:21 Completed GET /serviceworker.js 304 Not Modified in 2.999167ms
2021/06/20 21:20:33 Started GET / for [deleted]:0
2021/06/20 21:20:33 Started GET /css/index.css.map for [deleted]:0
2021/06/20 21:20:33 ...les/public/public.go:165:handle() [I] [Static] Serving /css/index.css.map
2021/06/20 21:20:33 Completed GET / 200 OK in 136.841325ms
2021/06/20 21:20:33 Completed GET /css/index.css.map 200 OK in 147.71492ms
2021/06/20 21:20:36 Completed GET /user/events 200 OK in 21.837993866s
2021/06/20 21:20:36 Started GET / for [deleted]:0
2021/06/20 21:20:36 Completed GET / 200 OK in 122.71882ms
2021/06/20 21:20:36 ...les/public/public.go:165:handle() [I] [Static] Serving /img/logo.svg
2021/06/20 21:20:36 ...les/public/public.go:165:handle() [I] [Static] Serving /css/index.css
2021/06/20 21:20:36 Started GET /css/index.css?v=7a6d6e9d8ba344b6a6460d087c47dbf9 for [deleted]:0
2021/06/20 21:20:36 Started GET /img/logo.svg for [deleted]:0
2021/06/20 21:20:36 Started GET /avatars/477ebe187c648228582162f32be00071?size=56 for [deleted]:0
2021/06/20 21:20:36 Started GET /avatars/477ebe187c648228582162f32be00071?size=48 for [deleted]:0
2021/06/20 21:20:36 Started GET /js/index.js?v=7a6d6e9d8ba344b6a6460d087c47dbf9 for [deleted]:0
2021/06/20 21:20:36 Completed GET /img/logo.svg 304 Not Modified in 3.155365ms
2021/06/20 21:20:37 ...les/public/public.go:165:handle() [I] [Static] Serving /js/index.js
2021/06/20 21:20:37 Completed GET /css/index.css?v=7a6d6e9d8ba344b6a6460d087c47dbf9 304 Not Modified in 7.348594ms
2021/06/20 21:20:37 Completed GET /avatars/477ebe187c648228582162f32be00071?size=48 304 Not Modified in 8.585939ms
2021/06/20 21:20:37 Completed GET /js/index.js?v=7a6d6e9d8ba344b6a6460d087c47dbf9 304 Not Modified in 9.77021ms
2021/06/20 21:20:37 Completed GET /avatars/477ebe187c648228582162f32be00071?size=56 304 Not Modified in 12.528699ms
2021/06/20 21:20:37 Started GET /api/v1/repos/search?sort=updated&order=desc&uid=1&team_id=0&q=&page=1&limit=15&mode=&archived=false for [2a03:580
2021/06/20 21:20:37 Started GET /api/v1/repos/search?sort=updated&order=desc&uid=1&team_id=0&q=&page=1&mode= for [2a03:5800:f8f8:e600:a5ce:2e22:23
2021/06/20 21:20:37 Completed GET /api/v1/repos/search?sort=updated&order=desc&uid=1&team_id=0&q=&page=1&mode= 200 OK in 50.877559ms
2021/06/20 21:20:37 Completed GET /api/v1/repos/search?sort=updated&order=desc&uid=1&team_id=0&q=&page=1&limit=15&mode=&archived=false 200 OK in 6
2021/06/20 21:20:37 Started GET /user/events for [deleted]:0
2021/06/20 21:20:37 Started GET /serviceworker.js for [deleted]:0
2021/06/20 21:20:37 ...les/public/public.go:165:handle() [I] [Static] Serving /serviceworker.js
2021/06/20 21:20:37 Completed GET /serviceworker.js 304 Not Modified in 1.767292ms
2021/06/20 21:20:39 Started GET /css/index.css.map for [deleted]:0


Reproxy-log:
2021/06/20 21:19:30.821 [WARN]  http: proxy error: context canceled
2021/06/20 21:20:00.846 [INFO]  request panic for /user/events from [deleted]:23419, net/http: abort Handler
2021/06/20 21:20:07.364 [WARN]  http: proxy error: net/http: timeout awaiting response headers
2021/06/20 21:20:07.458 [WARN]  http: proxy error: context canceled
2021/06/20 21:20:36.702 [INFO]  request panic for /user/events from [deleted]:23419, net/http: abort Handler
2021/06/20 21:20:47.718 [INFO]  request panic for /user/events from [deleted]:23419, net/http: abort Handler
2021/06/20 21:20:50.685 [INFO]  request panic for /user/events from [deleted]:23419, net/http: abort Handler
2021/06/20 21:20:55.720 [WARN]  http: proxy error: net/http: timeout awaiting response headers

**Comment by umputun** [comment]
I just tried the simplest gitea setup I can imagine, this is my compose

version: "3"

```yml
services:

  gitea:
    image: gitea/gitea:1.14.2
    container_name: gitea
    environment:
      - USER_UID=1001
      - USER_GID=1001
    restart: always
    volumes:
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
    ports:
      - "3000"
      - "2222:22"
    labels:
      reproxy.port: '3000'
      reproxy.route: '^/(.*)'
      reproxy.dest: '/$$1'
 
  reproxy:
    image: umputun/reproxy:master
    container_name: reproxy
    hostname: reproxy
    ports:
      - "80:8080"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
    environment:
      - TZ=America/Chicago
      - LISTEN=0.0.0.0:8080
      - DOCKER_ENABLED=true
      - DEBUG=true
```

worked just fine:

```
⟨bigmac@ umputun/tmp :|✔〕git clone http://localhost/umputun/test.git
Cloning into 'test'...
warning: You appear to have cloned an empty repository.
⟨bigmac@ umputun/tmp :|✔〕cd test
⟨bigmac@ tmp/test master|✔〕echo blah > 1.txt
⟨bigmac@ tmp/test master|…〕git add 1.txt
⟨bigmac@ tmp/test master|●1〕git commit -am "init"
[master (root-commit) 2b7ad31] init
 1 file changed, 1 insertion(+)
 create mode 100644 1.txt
⟨bigmac@ tmp/test master|✔〕git push
Username for 'http://localhost': umputun
Password for 'http://umputun@localhost':
Enumerating objects: 3, done.
Counting objects: 100% (3/3), done.
Writing objects: 100% (3/3), 201 bytes | 201.00 KiB/s, done.
Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
remote: . Processing 1 references
remote: Processed 1 references in total
To http://localhost/umputun/test.git
 * [new branch]      master -> master
⟨bigmac@ tmp/test master|✔〕
```
<img width="933" alt="rr346-202106-20143022-aph9d" src="https://user-images.githubusercontent.com/535880/122685919-16292800-d1d4-11eb-9c45-a5ef67f68838.png">

**Reply by saaremaa** [reply]
Ok. If you click on 2b7ad3152c link - you get page with diff?

**Reply by umputun** [reply]
yes, i do.

**Reply by umputun** [reply]
I made another commit directly from gitea, i.e. edited the file in gitea ui and committed it. no problem either

From the provided reproxy log fragment the only thing I can see is a timeout waiting for gitea's response. The [default values](https://github.com/umputun/reproxy/blob/master/app/main.go#L91) should be fine unless you modified them to some short timeouts.

**Reply by saaremaa** [reply]
OK. Thanks for the consultation. I will look for a problem from my part.
(Problem fixed): You were right with timeouts. After increasing the timeouts (read-header/resp-header) up to 8 seconds, the problem was corrected. Gitea + Reproxy is launched on the Raspberry PI3 , with high loaded other tasks. 

Thanks.
